Python/Module5_OddsAndEnds/Writing_Good_Code.ipynb

```diff
@@ -0,0 +1,1028 @@
+    "\n",
+    "```python\n",
+    "def get_first_and_last(x):\n",
+    "    \"\"\" Returns the first and last elements of `x`. `x` is \n",
+    "    assumed to be non-empty\n",
+    "    \"\"\" \n",
+    "    return (x[0], x[-1])\n",
+    "```\n",
+    "\n",
+    "Here are some examples of this function in action. Be sure that your type-hint captures all of this diversity.\n",
+    "\n",
+    "```python\n",
+    ">>> get_first_and_last(\"hello\")\n",
+    "('h', 'o')\n",
+    "\n",
+    ">>> get_first_and_last([0, 1, 2, 3])\n",
+    "(0, 3)\n",
+    "\n",
+    ">>> get_first_and_last((True, False))\n",
+    "(True, False)\n",
+    "```\n",
+    "\n",
+    "</div>"
+   ]
+  },
+  {
+   "cell_type": "markdown",
+   "metadata": {},
+   "source": [
+    "<div class=\"alert alert-info\">\n",
+    "\n",
+    "**Takeaway:**\n",
+    "\n",
+    "It is un-Pythonic to be overly strict when type hinting. Python handles typing flexibly - it uses [duck typing](https://en.wikipedia.org/wiki/Duck_typing) - and our type hints should follow suite. Rely on abstract base classes when possible to provide the appropriate guidelines for a particular variable type. \n",
+    "\n",
+    "</div>"
+   ]
+  },
+  {
+   "cell_type": "markdown",
+   "metadata": {},
+   "source": [
+    "## Documentation Styles\n",
+    "\n",
+    "[Davis King](https://twitter.com/nulhom?lang=en), a prolific and talented open source developer, is the creator of the [Dlib C++ library](http://dlib.net/). Among the library's major features, like machine learning algorithms, Davis lists its *documentation* as the first and foremost feature. In fact he says about Dlib:\n",
+    "\n",
+    "> *I consider the documentation to be the most important part of the library.* So if you find anything that isn't documented, isn't clear, or has out of date documentation, tell me and I will fix it.\n",
+    "\n",
+    "There is great wisdom in placing such a high value on documentation. In this section we will learn about two popular docstring-style specifications for Python: the NumPy specification and the Google specification. Both of these are rich extensions of the rudimentary docstring conventions that are proposed in [PEP 257](https://www.python.org/dev/peps/pep-0257/), and, critically, they both place an emphasis on documenting the variable types. \n",
+    "\n",
+    "PLYMI uses NumPy-style docstrings for throughout most of the text (except for when we were trying to keep the functions brief). This is ultimately just a choice of style/aesthetics. Ultimately, the critical takeaway here is to **pick a documentation style, learn it, and stick to it faithfully**. Once again, it is hard to overstate how important it is to anchor your code with clear and consistent documentation. It will aid you in your code-writing process, it will enable users to adopt and perhaps contribute to your code, and it will ensure longevity for your hard work."
```

Comment:
...uses NumPy-style docstrings throught most... <-- get rid of the for

...except for when were **are** trying to keep the functions brief <-- rather than were: "We use X when we are Y", not "we use X when we were Y"